src/reducers/vis-state-updaters.js

```diff
@@ -1231,65 +1239,156 @@ function closeSpecificMapAtIndex(state, action) {
  * @public
  */
 export const loadFilesUpdater = (state, action) => {
-  const {files, onFinish = loadFileSuccess} = action;
+  const {files, onFinish = loadFilesSuccess} = action;
   if (!files.length) {
     return state;
   }
 
-  const fileCache = [];
-  return withTask(
-    {
-      ...state,
-      fileLoading: true,
-      fileLoadingProgress: 0
-    },
-    makeLoadFileTask(files.length, files, fileCache, onFinish)
+  const fileLoadingProgress = files.reduce(
+    (accu, f, i) => merge_(initialFileLoadingProgress(f, i))(accu),
+    {}
   );
+  const fileLoading = {
+    fileCache: [],
+    filesToLoad: files,
+    onFinish
+  };
+  const nextState = merge_({fileLoading})(merge_({fileLoadingProgress})(state));
+
+  return loadNextFileUpdater(nextState);
 };
 
-export function loadNextFileUpdater(state, action) {
-  const {fileCache, filesToLoad, totalCount, onFinish} = action;
-  const fileLoadingProgress = ((totalCount - filesToLoad.length) / totalCount) * 100;
+// sucessfully loaded one file, move on to the next one
+export function loadFileStepSuccessUpdater(state, action) {
+  const {fileName, fileCache} = action;
+  const {filesToLoad, onFinish} = state.fileLoading;
+  const stateWithProgress = updateFileLoadingProgressUpdater(state, {
+    fileName,
+    progress: {percent: 1, message: 'Done'}
+  });
+
+  // save processed file to fileCache
+  const stateWithCache = pick_('fileLoading')(merge_({fileCache}))(stateWithProgress);
 
   return withTask(
-    {
-      ...state,
-      fileLoadingProgress
-    },
-    makeLoadFileTask(totalCount, filesToLoad, fileCache, onFinish)
+    stateWithCache,
+    DELAY_TASK(200).map(filesToLoad.length ? loadNextFile : () => onFinish(fileCache))
   );
 }
 
-export function makeLoadFileTask(totalCount, filesToLoad, fileCache, onFinish) {
+export function loadNextFileUpdater(state) {
+  const {filesToLoad} = state.fileLoading;
   const [file, ...remainingFilesToLoad] = filesToLoad;
 
+  // save filesToLoad to state
+  const nextStte = pick_('fileLoading')(merge_({filesToLoad: remainingFilesToLoad}))(state);
+
+  const stateWithProgress = updateFileLoadingProgressUpdater(nextStte, {
+    fileName: file.name,
+    progress: {percent: 0, message: 'loading...'}
+  });
+
+  return withTask(stateWithProgress, makeLoadFileTask(file, nextStte.fileLoading.fileCache));
+}
+
+export function makeLoadFileTask(file, fileCache) {
   return LOAD_FILE_TASK({file, fileCache}).bimap(
     // success
-    result =>
-      remainingFilesToLoad.length
-        ? loadNextFile({
-            fileCache: result,
-            filesToLoad: remainingFilesToLoad,
-            totalCount,
-            onFinish
+    gen =>
```

Comment:
That is a lot of newlines. Prettier is not kind to this type of code. Almost calls for a `// prettier-ignore`